src/type.c

```diff
@@ -30,13 +32,21 @@ SEXP vec_ptype(SEXP x, struct vctrs_arg* x_arg) {
   case vctrs_type_character:   return vec_ptype_slice(x, vctrs_shared_empty_chr);
   case vctrs_type_raw:         return vec_ptype_slice(x, vctrs_shared_empty_raw);
   case vctrs_type_list:        return vec_ptype_slice(x, vctrs_shared_empty_list);
-  case vctrs_type_dataframe:   return bare_df_map(x, &col_ptype);
+  case vctrs_type_dataframe:   return df_ptype(x, true);
   case vctrs_type_s3:          return s3_type(x, x_arg);
   case vctrs_type_scalar:      stop_scalar_type(x, x_arg);
   }
   never_reached("vec_ptype");
 }
 
+// [[ include("vctrs.h") ]]
+SEXP vec_ptype_unnamed(SEXP x, struct vctrs_arg* x_arg) {
```

Comment:
It does feel a bit awkward that we have to use `vec_ptype_unnamed()` in `vec_ptype2()` to get the unname behavior, but `vec_ptype()` retains names. I am also leaning towards letting `vec_ptype()` unname, which would remove the need for the specialized `vec_ptype_unnamed()`

Reply:
I feel like `vec_ptype()` always returning unnamed vectors would simplify the notion of a prototype in vctrs.